### ISSUE TITLE
Angular CLI build fails with allowJs: true fails when there is javascript file in src folder

### ISSUE THREAD

**Issue by yadue** [angular/angular#21080]
## I'm submitting a...

<pre><code>
[x] Regression (a behavior that used to work and stopped working in a new release)
</code></pre>

## Current behavior
ng serve fails if there is any javascript file in src directory

## Expected behavior
build should work just fine, last working version angular 4.x

## Minimal reproduction of the problem with instructions

create new project with `ng new projectname`
modify the `tsconfig.json` to set `"allowJs": true`
create new file .js file in src/app
run `ng serve`

## Environment

<pre><code>
Angular version: 5.1.1

Browser:
- [ x] Chrome (desktop) version XX
- [ ] Chrome (Android) version XX
- [ ] Chrome (iOS) version XX
- [ ] Firefox version XX
- [ ] Safari (desktop) version XX
- [ ] Safari (iOS) version XX
- [ ] IE version XX
- [ ] Edge version XX
 
For Tooling issues:
- Node version: 8.9.3
- Platform:  Windows 10 

Others:

@angular/cli 1.6.1
</code></pre>

it's not fixed by related bug https://github.com/angular/angular/issues/19757#issuecomment-352490677

**Comment by TerryHibbert** [comment]
I've got this issue with an imported JS file outside the src directory too:
`ERROR in error TS5055: Cannot write file '<JS FILE PATH>' because it would overwrite input file.`

**Comment by chuckjaz** [comment]
This looks like it might be an `ng` bug. I am having @hansl look into it.

**Comment by hansl** [comment]
I identified the error and looking at it (a PR will be pending soon and reference this).

This is not a bug in `ngc` but in the CLI. It's also not a regression since it's a problem with the new Angular 5 Webpack plugin, but Angular 4 will not have this problem (as it still uses the old compiler plugin).

**Comment by yadue** [comment]
Unforunatly I have new error, same scenario:
```
ERROR in ./src/main.ts
Module build failed: Error \test-app\src\main.ts is not part of the compilation output. Please check the other error messages for details.
    at AngularCompilerPlugin.getCompiledFile (\test-app\node_modules\@ngtools\webpack\src\angular_compiler_plugin.js:656:23)
    at plugin.done.then (\test-app\node_modules\@ngtools\webpack\src\loader.js:467:39)
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:188:7)
 @ multi webpack-dev-server/client?http://0.0.0.0:0 ./src/main.ts
ERROR in ./src/polyfills.ts
Module build failed: Error: \test-app\src\polyfills.ts is not part of the compilation output. Please check the other error messages for details.
    at AngularCompilerPlugin.getCompiledFile (\test-app\node_modules\@ngtools\webpack\src\angular_compiler_plugin.js:656:23)
    at plugin.done.then (\test-app\node_modules\@ngtools\webpack\src\loader.js:467:39)
    at <anonymous>
    at process._tickCallback
```

it raises exactly when this line is removed:
`this._compilerOptions.outDir = '';`

scenario the same:
create new project with ng new projectname
modify the tsconfig.json to set "allowJs": true
create new file .js file in src/app
run ng serve

**Comment by yadue** [comment]
Seems to be fine now for simple scenario, additionaly same error happens when:
ng serve --aot
ng build --aot

**Comment by neoncube2** [comment]
Thanks for opening this separate ticket, @yadue :)

I'm seeing this error message using Angular 5.1.2, just FYI.

**Comment by neoncube2** [comment]
Okay, so, I'm seeing pretty strange (but helpful!) behaviour with this.

For example, if I have the following line and compile with --watch=auto, I get the compile errors that are mentioned in this bug report.
`import * as CSVParser from '../../../../../node_modules/js-csvparser';`

If I comment out the line, it automatically recompiles (since I'm using --watch=auto), and then everything is fine. However, the next time that I recompile, I get the errors again, but uncommenting and then commenting out that line makes them go away again.

Basically: uncommenting and commenting the problematic line seems to make the error go away when using --watch=auto.

**Comment by neoncube2** [comment]
Possible workaround:
https://github.com/angular/angular/issues/19757#issuecomment-353933280

**Comment by itsnotvalid** [comment]
@neoncube2 this may not really help, as the critical part is that `--prod` with `allowJs` == `true` needs to work.

**Comment by neoncube2** [comment]
@itsnotvalid The workaround works for me when using `--prod` (and `allowJs: true`) as well :)

Just make sure that you have `--watch=auto` set and then make a change after the first build fails.

**Comment by itsnotvalid** [comment]
@neoncube2 then what you are saying is that the output of `--watch=auto` is inconsistent to the one without, which is a bigger issue.

**Comment by yadue** [comment]
But how to run it when using CI like jenkins? Makes no sense to me.

**Comment by jackjamieson** [comment]
@neoncube2 not really a workaround like @yadue  said it needs to work without manual intervention.

**Comment by rctyler** [comment]
I am also seeing this issue on angular v5.0.5 and angular-cli v1.5.5

**Comment by neoncube2** [comment]
@yadue and @jackjamieson, Yes, this required manual intervention, so it probably won't when building via CI. Personally, I'm building via the CLI and then just FTP'ing the built files to my server, so the workaround works for me.

@itsnotvalid, I agree that the output of `--auto` being different than without sounds like something that maybe someone with good knowledge of the way Angular builds should look at.

**Comment by florinbardosi** [comment]
There is a plan to have also this issue fixed and released? We waited #19757 to get fixed with the idea that it will fix our issue.

**Comment by DragoTechInnovations** [comment]
Any idea as to when this issue will be resolved?
I have to update my project to material 5 which depends on ng5 which fails on build with `allowJs`.

**Comment by samanmohamadi** [comment]
A temporary workaround could be this:
create a node package with your js file and add to your project as a dependency.
 [install-local](https://www.npmjs.com/package/install-local) is useful if you want to install this package locally.
there is no need to `allowJs` anymore.

**Comment by DragoTechInnovations** [comment]
@samanmohamadi what if there are no packages for those files? Like chartist plugins (my use case)

**Comment by samanmohamadi** [comment]
1. create a directory (ex chartist-plugins )
2. copy your js file into chartist-plugins directory
3. ` cd path/to/chartist-plugins`
4.  run `npm init` (just enter for all questions)
5. `cd path/to/your/angular/project`
6.  `install-local /path/to/the/created/directory` (assumed you have already installed[ install-local](https://www.npmjs.com/package/install-local))
7. import the package like other npm packages in your project
8. remove **allowJs** from tsconfig

**Comment by NateWarner** [comment]
Appreciate the workaround. Just want to pipe in that this is affecting us as well.

**Comment by vladimiry** [comment]
Another workaround is:
- defining similar path alias in `src/tsconfig.app.json`:
```json
    "baseUrl": ".",
    "paths": {
      "_raw-js/*": [
        "./raw-js/*"
      ]
    },
```
- defining declaration stub for that location, `src/typings.d.ts` file:
```typescript
declare module "_raw-js/*";
```
- and then importing it like this:
```typescript
const {plot} = require("_raw-js/stats/plot");
```

**Comment by gregdbqn** [comment]
we also have the issue, and packaging the library in the a npm didn't work for that particular third party library. 
I'm going to go back to Angular 4.x meanwhile

**Comment by ameya** [comment]
Anyone has a workaround for Angular 5.
Install local does not work on my windows machine.  [link](https://github.com/nicojs/node-install-local/issues/5).
@vladimiry workaround also did not work out. 
@hansl is there any ETA on this PR it has been a blocker in my project.

**Comment by vladimiry** [comment]
> @vladimiry workaround also did not work out.

@ameya try to execute `npm i --save-dev @types/node` first of all.

**Comment by NITESH11DEVIL** [comment]
Hi @hansl  we are also facing the same issue is there any ETA on this 
resolution.

**Comment by gregdbqn** [comment]
downgrading to Angular 4.4.6 solved the problem for me. I understand it's not a solution for everyone though

**Comment by ameya** [comment]
Hi @hansl, is there any update on this issue is resolved? As we already have a production-ready application build from angular 5 downgrading to angular 4 is not an option for us. Because of this one blocker, we are thinking of breaking only the external js dependant part into angular 4 application that will be used in angular 5 application via html iFrames. Any ETA date so we can hold on to bug to avoid unnecessary extra work?

**Comment by itsnotvalid** [comment]
@gregdbqn not an option when you want to use material for angular.
I've been tracking this issue since Nov from #19757 and it has been at least a month for this issue. I know it's an open-source project so everyone should be contributing, but the complexity of the code in this part in angular is also very high.

**Comment by vladimiry** [comment]
Btw going without CLI and respectively without Angular compiler might be an option. I don't mean using ejected from CLI Weback config, but custom Webpack config containing for example stuff like `awesome-typescript-loader`, `angular2-template-loader`, `angular-router-loader` etc.

**Comment by neoncube2** [comment]
@ameya, Did you try the workaround that I posted?

**Comment by ameya** [comment]
@neoncube2 you solution to use --prod and allowJs in tsconfig.json did not work.
Let me know if I am doing anything wrong here.

```
D:\myApp\myAppTest\>ng build --prod
Date: 2018-01-31T03:08:09.794Z
Hash: 92dfe56172b695fee9b7
Time: 62063ms
chunk {scripts} scripts.162ac3c01fde25d9d8ce.bundle.js (scripts) 196 kB [initial] [rendered]
chunk {0} styles.27f90adbbbe0102a4e65.bundle.css (styles) 249 kB [initial] [rendered]
chunk {1} polyfills.3bc34265385d52184eab.bundle.js (polyfills) 86 bytes [initial] [rendered]
chunk {2} main.e402deade8b026b7d50e.bundle.js (main) 84 bytes [initial] [rendered]
chunk {3} inline.d5e8e847c6fbdb00d982.bundle.js (inline) 1.45 kB [entry] [rendered]

ERROR in error TS5055: Cannot write file 'D:/iCTES/angular5/ui-modelling/myAppTest\/node_modules/jsplumb/dist/js/jsplumb.js' because it would overwrite input file.
  Adding a tsconfig.json file will help organize projects that contain both TypeScript and JavaScript files. Learn more at https://aka.ms/tsconfig.

D:\myApp\myAppTest\>
```

My tsconfig.json
```

{
  "compileOnSave": false,
  "compilerOptions": {
    "allowJs":true,
    "outDir": "./dist/out-tsc",
    "sourceMap": true,
    "declaration": false,
    "moduleResolution": "node",
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "target": "es5",
    "typeRoots": [
      "node_modules/@types"
    ],
    "lib": [
      "es2017",
      "dom"
    ]
  }
}
```

**Comment by neoncube2** [comment]
@ameya, you also need to use `--watch=auto`, so the line would be something like this: `ng build --prod --watch=auto`

After running the command, you should receive an error, like the one that you posted. However, if you then go into one of the .ts files that's in the project, make a change, and then save the file, the app should automatically rebuild without the error.

Hope that helps! :)  If it still doesn't work for some reason, please let me know.

**Comment by ameya** [comment]
Thanks @neoncube2 , But it did not help.

```
D:\myApp\myAppTest>ng build --prod --watch=auto
Date: 2018-01-31T04:00:01.925Z
Hash: 92dfe56172b695fee9b7
Time: 48916ms
chunk {scripts} scripts.162ac3c01fde25d9d8ce.bundle.js (scripts) 196 kB [initial] [rendered]
chunk {0} styles.27f90adbbbe0102a4e65.bundle.css (styles) 249 kB [initial] [rendered]
chunk {1} polyfills.3bc34265385d52184eab.bundle.js (polyfills) 86 bytes [initial] [rendered]
chunk {2} main.e402deade8b026b7d50e.bundle.js (main) 84 bytes [initial] [rendered]
chunk {3} inline.d5e8e847c6fbdb00d982.bundle.js (inline) 1.45 kB [entry] [rendered]

ERROR in error TS5055: Cannot write file 'D:/myApp/myAppTest/node_modules/jsplumb/dist/js/jsplumb.js' because it would overwrite input file.
  Adding a tsconfig.json file will help organize projects that contain both TypeScript and JavaScript files. Learn more at https://aka.ms/tsconfig.
```

**Comment by elvisfernandes** [comment]
@ameya after you get this error you must save any .ts file to start the transpile process again, then you'll get no error.

**Comment by ameya** [comment]
@elvisfernandes thanks for the help we can work with the workaround now. I missed out the purpose of --watch=auto by @neoncube2. Thanks for the help  :)

**Comment by ameya** [comment]
I have tested out the newer version of cli that is 1.6.8 but still, the issue is not resolved. Followed the steps mentioned in [this link](https://stackoverflow.com/questions/43931986/how-to-upgrade-angular-cli-to-the-latest-version) to upgrade

**Comment by trotyl** [comment]
@ameya The first version including this fix is 1.7.0-rc.0, you can just check the CLI releases log.

**Comment by ameya** [comment]
@trotyl how do we add 1.7.0-rc.0 simply changing the package.json cli version to 1.7.0-rc.0 followed by npm install? Because with npm install --save-dev @angular/cli@latest I am getting only 1.6.8. Adding @shashankgautam1811 with a similar issue.

**Comment by trotyl** [comment]
@ameya It's NPM support question, should not be made here. Beta versions are always marked `next`, not `latest`.

And `1.7.0` has already been released now.

**Comment by mlzarathustra** [comment]
I'm having the same problem with Angular 1.7.4

I dropped back into JS because TS doesn't seem to support spread array construction.
